[SVG] Move marker related code from LegacyRenderSVGShape to LegacyRenderSVGPath

Constellation opened the pull request:
#### 95664cfd77a025b5154b8fc6dd52a6cbd9c0f544
<pre>
[SVG] Move marker related code from LegacyRenderSVGShape to LegacyRenderSVGPath
<a href="https://bugs.webkit.org/show_bug.cgi?id=262756">https://bugs.webkit.org/show_bug.cgi?id=262756</a>
rdar://116557017

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink&apos;s work[1].

This change moves marker related code from LegacyRenderSVGShape to LegacyRenderSVGPath
because only LegacyRenderSVGPath can have markers. `&lt;path&gt;`, `&lt;line&gt;`, `&lt;polyline&gt;` and `&lt;polygon&gt;`
can have that and all creates LegacyRenderSVGPath.

1. This makes LegacyRenderSVGShape::updateShapeFromElement more concise. This now only creates path, and computes two bounding boxes.
2. Previously we were calling LegacyRenderSVGShape::updateShapeFromElement from various places to update path. But this is not a good
   way since it updates m_fillBoundingBox and m_strokeBoundingBox while the class&apos; updateShapeFromElement already filled that.
   Since now LegacyRenderSVGShape::updateShapeFromElement is just creating a path and computing these two bounding boxes, we introduce
   ensurePath and use it in the places which is not overriding LegacyRenderSVGShape::updateShapeFromElement. This avoids accidentally
   updating already computed m_fillBoundingBox / m_strokeBoundingBox computing again. If they gets different, then it becomes a bug.
   (If newly path-based computed m_strokeBoundingBox is different from m_strokeBoundingBox computed in the class&apos;s updateShapeFromElement,
   getBBox etc. can return different value at different timing). This change makes m_strokeBoundingBox and m_fillBoundingBox more consistent.
3. We duplicate the change in LBSE.

[1]: <a href="https://bugs.chromium.org/p/chromium/issues/detail?id=435097">https://bugs.chromium.org/p/chromium/issues/detail?id=435097</a>

* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
(WebCore::RenderSVGEllipse::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
(WebCore::RenderSVGPath::calculateUpdatedStrokeBoundingBox const):
(WebCore::markerForType):
(WebCore::RenderSVGPath::shouldGenerateMarkerPositions const):
(WebCore::RenderSVGPath::drawMarkers):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
(WebCore::RenderSVGPath::processMarkerPositions):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
(WebCore::RenderSVGRect::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::updateShapeFromElement):
(WebCore::RenderSVGShape::fillStrokeMarkers):
(WebCore::RenderSVGShape::ensurePath):
(WebCore::RenderSVGShape::shouldGenerateMarkerPositions const): Deleted.
(WebCore::markerForType): Deleted.
(WebCore::RenderSVGShape::computeMarkerBoundingBox const): Deleted.
(WebCore::RenderSVGShape::drawMarkers): Deleted.
(WebCore::RenderSVGShape::processMarkerPositions): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::shapeType const):
(WebCore::RenderSVGShape::drawMarkers):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
(WebCore::LegacyRenderSVGEllipse::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
(WebCore::LegacyRenderSVGPath::calculateUpdatedStrokeBoundingBox const):
(WebCore::markerForType):
(WebCore::LegacyRenderSVGPath::shouldGenerateMarkerPositions const):
(WebCore::LegacyRenderSVGPath::drawMarkers):
(WebCore::LegacyRenderSVGPath::markerRect const):
(WebCore::LegacyRenderSVGPath::processMarkerPositions):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
(WebCore::LegacyRenderSVGRect::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::updateShapeFromElement):
(WebCore::LegacyRenderSVGShape::layout):
(WebCore::LegacyRenderSVGShape::fillStrokeMarkers):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::ensurePath):
(WebCore::LegacyRenderSVGShape::shouldGenerateMarkerPositions const): Deleted.
(WebCore::markerForType): Deleted.
(WebCore::LegacyRenderSVGShape::markerRect const): Deleted.
(WebCore::LegacyRenderSVGShape::drawMarkers): Deleted.
(WebCore::LegacyRenderSVGShape::processMarkerPositions): Deleted.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::drawMarkers):

Canonical link: <a href="https://commits.webkit.org/269245@main">https://commits.webkit.org/269245@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e99d3d83fe396c476ba0ac9959dcc514ec151ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23886 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20371 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22248 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26466 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22532 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22233 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/21887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24738 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/19000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/19944 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24070 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20595 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19949 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24154 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2735 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20546 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
```diff
@@ -123,8 +123,8 @@ bool LegacyRenderSVGEllipse::shapeDependentStrokeContains(const FloatPoint& poin
     // The optimized code below does not support non-smooth strokes so we need to
     // fall back to LegacyRenderSVGShape::shapeDependentStrokeContains in these cases.
-    if (!hasSmoothStroke() && !hasPath())
-        LegacyRenderSVGShape::updateShapeFromElement();
+    if (!hasSmoothStroke())
+        ensurePath();
```
Who is going to calculate the `m_fillBoundingBox` and `m_strokeBoundingBox`? I think is a weird state of the renderer. The path is created from the element but the bounding boxes are empty.

Constellation replied:
`m_fillBoundingBox` and `m_strokeBoundingBox` computation is the responsibility of `updateShapeFromElement`, invoked by `layout`, not in this function. So when calling `shapeDependentStrokeContains`, it already has these information computed from `updateShapeFromElement`, this is an invariant of this class.

For example, in this function, we are using `m_center` in `!hasPath()` case, and this `m_center` is computed  in `updateShapeFromElement` (which is calling `calculateRadiiAndCenter` to compute it), but this function itself is not calling that. This is because `updateShapeFromElement ` is the function which updates these values (including `m_strokeBoundingBox` etc.), and it is already called by `layout` when these values need to be updated.

Another example. In `LegacyRenderSVGShape::strokeContains`, it is already touching `m_strokeBoundingBox` via `strokeBoundingBox()`. This is because this function is relying on `updateShapeFromElement`, which is called by `layout`.

When something affecting on these things gets changed and if we would like to invoke `shapeDependentStrokeContains`, then we need to update layout information to reflect all related styles, attributes, etc., and `updateShapeFromElement` does `m_fillBoundingBox` and `m_strokeBoundingBox` update for that.

We are not always computing these values from the path since some of shape can compute them without having a path. `LegacyRenderSVGEllipse::updateShapeFromElement` is one of the example, which is computing `m_fillBoundingBox` and `m_strokeBoundingBox` without creating a path.

The reason why we need to have a path here is that we need path based approach for non-smooth-stroke and use `LegacyRenderSVGShape::shapeDependentStrokeContains`. Our current code is intentionally not creating `m_path` for non-smooth-stroke in `LegacyRenderSVGEllipse::updateShapeFromElement` because the goal of that function is computing `m_fillBoundingBox` and `m_strokeBoundingBox`, and that does not require a path for non-smooth-stroke (so our current code has optimization for that). And only this case needs a path, but bounding boxes are already computed correctly by `LegacyRenderSVGEllipse::updateShapeFromElement` (And it is not `shapeDependentStrokeContains`'s responsibility).

So, this patch makes it more consistent behavior. Here we only materialize a path because it is necessary for stroke-based contains operations in `LegacyRenderSVGShape::shapeDependentStrokeContains`. But we do not need `m_fillBoundingBox` and `m_strokeBoundingBox`. They are computed in `LegacyRenderSVGEllipse::updateShapeFromElement`, which is already invoked by `layout`.

heycam replied:
After this patch there will be no explicit calls to `LegacyRenderSVGShape::updateShapeFromElement` from functions other than `updateShapeFromElement` overrides, is that right? I think that's a positive change since we'll no longer be using it for two purposes (i.e. the main purpose of updating shape data from layout, and this secondary purpose of ensuring the Path exists in functions like this one that happen to need it).

Constellation replied:
Right! That's the main purpose of this patch. After this patch, we no longer call `LegacyRenderSVGShape::updateShapeFromElement` from non `updateShapeFromElement` places. That makes these responsibility, updating m_strokeBoundingBox and m_fillBoundingBox, explicit and clear.

This patch paves a way to lazily compute `m_strokeBoundingBox` by making `updateShapeFromElement` more and more light weight (this is the next patch), and the final patch will start avoiding `m_strokeBoundingBox` until it becomes really necessary.

Also, that change will effectively make `LegacyRenderSVGShape::updateShapeFromElement` just `m_fillBoundingBox = ensurePath().boundingBox();`. Probably I'll make `LegacyRenderSVGShape::updateShapeFromElement` as pure virtual function, and moving them to each derived `updateShapeFromElement` function (because it now just becomes `m_fillBoundingBox = ensurePath().boundingBox();`, then each derived shape's `updateShapeFromElement` becomes really concise, easy to read, and clear.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -50,9 +50,10 @@ LegacyRenderSVGPath::~LegacyRenderSVGPath() = default;
     updateZeroLengthSubpaths();
 
-    m_strokeBoundingBox = calculateUpdatedStrokeBoundingBox();
+    m_strokeBoundingBox = calculateUpdatedStrokeBoundingBox(m_strokeBoundingBox);
```
What is point in passing this rectangle to another method in the same class? I do not see any other calls to this method.

Constellation replied:
This change makes `calculateUpdatedStrokeBoundingBox` usable for the specified input. And we are now using `m_strokeBoundingBox`. But we use this for approximate stroke bounding box to attach markers etc. rect to that.

heycam replied:
By this do you mean that in a future patch we might pass in different values to this function, and it won't always operate on `m_strokeBoundingBox`?

Existing code, but I find the name `calculateUpdatedStrokeBoundingBox` a bit unclear. "updated" is very generic. Seems like the purpose of this function is to adjust the already computed stroke bounding box by merging it with (a) the marker rectangles and (b) the zero length path segment linecap rectangles. What do you think about renaming it to something like `adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps`? Possibly operating on `FloatRect&` instead of returning a new value.

Constellation replied:
> By this do you mean that in a future patch we might pass in different values to this function, and it won't always operate on m_strokeBoundingBox?

Right. It will get approximate repaint bounding box next.

> Existing code, but I find the name calculateUpdatedStrokeBoundingBox a bit unclear. "updated" is very generic. Seems like the purpose of this function is to adjust the already computed stroke bounding box by merging it with (a) the marker rectangles and (b) the zero length path segment linecap rectangles. What do you think about renaming it to something like adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps? Possibly operating on FloatRect& instead of returning a new value.

Sounds good! That makes the function clear.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -63,9 +64,10 @@ void LegacyRenderSVGPath::updateShapeFromElement()
 {
-    FloatRect strokeBoundingBox = m_strokeBoundingBox;
+    if (!m_markerPositions.isEmpty())
+        strokeBoundingBox.unite(markerRect(strokeWidth()));
```
The code below gets the `strokeWidth()` in a local variable `strokeWidth`. I think we should be consistent and avoid calling this method twice.

Constellation replied:
Changed.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
```diff
@@ -138,8 +138,8 @@ bool LegacyRenderSVGRect::shapeDependentStrokeContains(const FloatPoint& point,
 {
     // The optimized code below does not support non-smooth strokes so we need to
     // fall back to LegacyRenderSVGShape::shapeDependentStrokeContains in these cases.
-    if (!hasSmoothStroke() && !hasPath())
-        LegacyRenderSVGShape::updateShapeFromElement();
+    if (!hasSmoothStroke())
+        ensurePath();
```
Ditto.

Constellation replied:
Ditto. Updating this is `updateShapeFromElement`'s purpose.

shallawa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -274,11 +258,18 @@ void LegacyRenderSVGShape::fillStrokeMarkers(PaintInfo& childPaintInfo)
+            switch (m_shapeType) {
+            case ShapeType::Line:
+            case ShapeType::Path:
+                downcast<LegacyRenderSVGPath>(*this).drawMarkers(childPaintInfo);
```
Why do we have to cast here from `LegacyRenderSVGShape` to `LegacyRenderSVGPath`? Can't we use kind of polymorphism to avoid this casting?

Constellation replied:
I added this not to do much virtual function calls when we already know what class we need to dispatch. But I'm totally ok to have `drawMarkers` virtual function here since I know this part is not getting perf concerns from the benchmark. Changed.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGShape.cpp:
```diff
@@ -149,6 +147,7 @@ void RenderSVGShape::layout()
     if (m_needsShapeUpdate) {
         // FIXME: [LBSE] Upstream SVGLengthContext changes
         // graphicsElement().updateLengthContext();
+        m_shapeType = ShapeType::Empty;
```
This is a weird place to make this resetting. I think this member has to be initialized with `ShapeType::Empty`. It has be set in `updateShapeFromElement()`. And it can be reset to `ShapeType::Empty ` in `clearPath()`.

Constellation replied:
Yeah, let's put this in derived class' `updateShapeFromElement()`. So we can clear this state first to empty, and using early return when it is actually empty (for example, empty rect needs to be empty state).
We do not want to put `Empty` in `LegacyRenderSVGShape::updateShapeFromElement` since it is called from the derived class, and it clears the type to `Empty` incorrectly if we do that. Probably we can remove `LegacyRenderSVGShape::updateShapeFromElement` after strokeBoundingBox computation is purged from that (we will do it lazily to avoid that when it is not necessary, as the same to path materialization), so `LegacyRenderSVGShape::updateShapeFromElement` will become just `m_fillBoundingBox = ensurePath().boundingRect()`. Changed.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGRect.cpp:
```diff
@@ -67,7 +67,11 @@ void RenderSVGRect::updateShapeFromElement()
-    if (rectElement().rx().value(lengthContext) > 0 || rectElement().ry().value(lengthContext) > 0 || hasNonScalingStroke()) {
+    m_shapeType = ShapeType::Rectangle;
+    if (rectElement().rx().value(lengthContext) > 0 || rectElement().ry().value(lengthContext) > 0)
+        m_shapeType = ShapeType::RoundedRectangle;
```
Using if-else will be clearer.

Constellation replied:
Changed.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
```diff
@@ -62,6 +62,11 @@ void RenderSVGEllipse::updateShapeFromElement()
     if (m_radii.isEmpty())
         return;
 
+    if (m_radii.width() == m_radii.height())
+        m_shapeType = ShapeType::Circle;
+    else
+        m_shapeType = ShapeType::Ellipse;
```
This `shapeType` is really weird. I looked at the complete patch in https://github.com/WebKit/WebKit/pull/18444 and I did not find any difference in treating the `Circle` from the `Ellipse`. They are always treated the same way. So why we have to differentiate between them here? And why do we even bother by having `m_shapeType`?

Can't we have a virtual method like `RenderSVGShape::canUseStrokeApproximation()` `RenderSVGShape::isSimpleStroke()` or `RenderSVGShape::hasSimpleStroke()` with the default value false, and make `RenderSVGEllipse` and `RenderSVGRect` return true for it?

Constellation replied:
This is because the shapeType for Ellipse and Circle is the critical part of the hit testing optimization. Right now we are always asking to path when there is a path. But if the stroke is smooth and it is a circle, we do not need to use path-based hit-testing (we can see whether the hit-testing-point from the center of the circle is within the stroke-width of the circle).

shallawa replied:
My question is why do not we have `ShapeType::Ellipse` and get rid of `ShapeType::Circle'? So this code can be 
```
m_shapeType = ShapeType::Ellipse;
```
Alternatively we define this function 
```
class RenderSVGShape : public RenderSVGModelObject {
     virtual bool canUseStrokeApproximation() { return false; }
};
```
And override it for RenderSVGEllipse:
```
class RenderSVGEllipse final : public RenderSVGShape  {
     bool canUseStrokeApproximation() override { return true; }
};
```

Constellation replied:
For `canUseStrokeApproximation`, this is not necessary to make it a virtual function because `shapeDependentStrokeContains` is already a virtual function. We can define this function in the derived class without virtual since we will have the fast path in each derived class.
We will define `bool RenderSVGEllipse::canUseStrokeApproximation() const` and use it in `RenderSVGEllipse::shapeDependentStrokeContains`.

And when using `ShapeType` etc. for the fast path, it is not approximation since this is not a repaint. So probably we will define it like, `canUseStrokeContainsFastPath`

And when defining this function, it is useful to have this `ShapeType::Circle` because this tells quickly that we can use the optimization or not.

So, then inside `RenderSVGEllipse::shapeDependentStrokeContains`

```
if (!canUseStrokeContainsFastPath()) {
    ensurePath();
    return RenderSVGShape::shapeDependentStrokeContains(...);
}

... doing the fast path ...
```

And then, implementation of `canUseStrokeContainsFastPath` will be,

```
bool RenderSVGEllipse::canUseStrokeContainsFastPath() const  // Not virtual function.
{
    if (hasNonScalingStroke())
        return false;
    return m_shapeType == ShapeType::Circle && style().svgStyle().strokeDashArray().isEmpty();
}
```

And we will use `ShapeType::Circle`. This circle shape-type makes it clear the intent of this fast path. When it is circle and it is not a dash stroke, then we can compute the fast path.

Then, why we cannot use the fast path with Ellipse too and just always having Ellipse instead of adding Circle is that the current fast path has an issue, and this one should work only with Circle case. This is https://bugs.chromium.org/p/chromium/issues/detail?id=393143, and the hittest is included in WPT, and we are right now failing because we are using fast path for Ellipse case too. https://wpt.fyi/results/svg/interact/scripted/ellipse-hittest.html?label=experimental&label=master&aligned
The reason why the current fast path does not work is that we are assuming that half-stroke-width is enough to compute hit-test with radius from the center-point. But this only works with Circle and this does not work with Ellipse.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGShape.cpp:
```diff
@@ -64,8 +64,6 @@ RenderSVGShape::~RenderSVGShape() = default;
 void RenderSVGShape::updateShapeFromElement()
 {
     m_path = createPath();
```
Should not this be `ensurePath()`?

Constellation replied:
This needs to be `createPath()` since this function is updating the path with the current state. The `m_path` in this class is the stale path which is created for the previous state, and we need to update with the new one.
In the other `updateShapeFromElement` functions, it is calling `clearPath()` to nullify it.
This place's intent includes `clearPath()`: This is effectively,

```
clearPath();
ensurePath();
```

shallawa commented on Source/WebCore/rendering/svg/RenderSVGPath.cpp:
```diff
@@ -52,15 +52,22 @@ RenderSVGPath::~RenderSVGPath() = default;
     updateZeroLengthSubpaths();
 
-    m_strokeBoundingBox = calculateUpdatedStrokeBoundingBox();
+    m_strokeBoundingBox = calculateUpdatedStrokeBoundingBox(m_strokeBoundingBox);
```
I think the life cycle of the `m_path` and the flags `m_needsShapeUpdate` and `m_needsBoundariesUpdate` became more unclear and fragile with this patch. I think after creating `m_path` and updating it from the element, `m_needsShapeUpdate` should be set to false. And I think `m_strokeBoundingBox`, `m_fillBoundingBox` and `m_repaintBoundingBox` should be calculated in something called `updateBoundariesBoxes()` and after calling it, `m_needsBoundariesUpdate` should be set to false.

shallawa replied:
Please have a look at https://github.com/WebKit/WebKit/pull/18490. This crash is happening because `m_path` is null even after running the layout. This shows that this code is fragile and there are cases that we are not aware of and they are hard to find.

Constellation replied:
I see #18490, and as the bug is showing, this is not having `m_path` even without this patch's change. (so existing bug). I'll comment on that issue about it.

The reason of that bug is not directly related to this change since,

1. The bug is anyway happening without the change.
2. The bug is that LegacyRenderSVGPath is not having a `m_path`. Since that class always has a path when `updateShapeFromElement` happens, this means that it is not doing layout before, while `isRenderingDisabled` code in LegacyRenderSVGPath is requiring that.
3. Stacktrace of #18490 is showing that `LegacyRenderSVGRoot::layout` is called for the container, and it is calling `SVGRenderSupport::layoutChildren`. This means that the bug is happening because of inconsistency between `isRenderingDisabled` condition and layout-skipping condition in `SVGRenderSupport::layoutChildren`. If we align them, then it is fine. If we skipped the layout (this is what is happening in that bug), then layout related condition is not usable, including `m_path`. Using layout dependent condition in `isRenderingDisabled` is not right.

I think that calling `updateShapeFromElement` from various places makes lifecycle tracking harder, as the above issue is showing the crash. The better direction is that making it much more explicit and clear (Always update from `updateShapeFromElement` calls by layout)

So,
> This crash is happening because m_path is null even after running the layout. 

This is not the reason of the crash in #18490. The root cause of the bug is, layout is not properly happening for that LegacyRenderSVGPath, and/or not using the right condition in `isRenderingDisabled`.
It is skipping layout because of the condition in `SVGRenderSupport::layoutChildren`, and not using the same condition in `isRenderingDisabled`. That's the issue. If we skip the layout with a condition, then we need to check the same condition again. `isRenderingDisabled` works only if the layout is once called. If the layout is skipped, using it is not the right way.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
```diff
@@ -128,18 +130,15 @@ class LegacyRenderSVGShape : public LegacyRenderSVGModelObject {
     void fillStrokeMarkers(PaintInfo&);
-    void drawMarkers(PaintInfo&);
+
+    virtual void drawMarkers(PaintInfo&) { };
```
Nit: semicolon not needed.

Constellation replied:
Oops, removed.

heycam commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -49,27 +49,31 @@ LegacyRenderSVGPath::~LegacyRenderSVGPath() = default;
 
     ASSERT(hasPath());
     if (path().isEmpty())
-        m_shapeType = ShapeType::Empty;
```
Is there a particular reason to move this to the top of the function? It looks like we don't need to do this, and I think it is clearer to have it here with the other assignments to `m_shapeType`. Or is that you want to consistently start every `updateShapeFromElement` function with `m_shapeType = ShapeType::Empty`?

Constellation replied:
> Or is that you want to consistently start every updateShapeFromElement function with m_shapeType = ShapeType::Empty?

Yes, this is the reason why I putted this on the first line. Previously I was reseting it in the `layout` function side before calling `updateShapeFromElement`, but I changed it from https://github.com/WebKit/WebKit/pull/18740#discussion_r1353472909 's comment.
However, the largest concern to me was missing Empty clearing because I once had that issue when creating a prototype in one derived class, and it caused some test failures (while creating a prototype patch). So not to repeat that, I pedantically aligned the prologue of these functions, starting with clearing of the `m_shapeType`.

heycam reviewed (approved):
As an overall comment, I think it is difficult to understand when various bits of state on these classes is updated, and who is responsible for doing it, without doing a lot of reading. I wonder if some better function names would help, but I don't have any concrete suggestions right now.

It also makes me wonder whether these classes would be easier to understand if we treated all of { path-or-shape-values, path, fill bbox, stroke bbox } as lazily computable states, instead of relying on "knowing" that certain of them are computed at certain times.

Constellation committed:
[SVG] Move marker related code from LegacyRenderSVGShape to LegacyRenderSVGPath
https://bugs.webkit.org/show_bug.cgi?id=262756
rdar://116557017

Reviewed by Cameron McCormack.

This is one step of the patch series implementing approximate repainting rect for SVG path, derived from blink's work[1].

This change moves marker related code from LegacyRenderSVGShape to LegacyRenderSVGPath
because only LegacyRenderSVGPath can have markers. `<path>`, `<line>`, `<polyline>` and `<polygon>`
can have that and all creates LegacyRenderSVGPath.

1. This makes LegacyRenderSVGShape::updateShapeFromElement more concise. This now only creates path, and computes two bounding boxes.
2. Previously we were calling LegacyRenderSVGShape::updateShapeFromElement from various places to update path. But this is not a good
   way since it updates m_fillBoundingBox and m_strokeBoundingBox while the class' updateShapeFromElement already filled that.
   Since now LegacyRenderSVGShape::updateShapeFromElement is just creating a path and computing these two bounding boxes, we introduce
   ensurePath and use it in the places which is not overriding LegacyRenderSVGShape::updateShapeFromElement. This avoids accidentally
   updating already computed m_fillBoundingBox / m_strokeBoundingBox computing again. If they gets different, then it becomes a bug.
   (If newly path-based computed m_strokeBoundingBox is different from m_strokeBoundingBox computed in the class's updateShapeFromElement,
   getBBox etc. can return different value at different timing). This change makes m_strokeBoundingBox and m_fillBoundingBox more consistent.
3. We duplicate the change in LBSE.

[1]: https://bugs.chromium.org/p/chromium/issues/detail?id=435097

* Source/WebCore/rendering/svg/RenderSVGEllipse.cpp:
(WebCore::RenderSVGEllipse::updateShapeFromElement):
(WebCore::RenderSVGEllipse::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::updateShapeFromElement):
(WebCore::RenderSVGPath::calculateUpdatedStrokeBoundingBox const):
(WebCore::markerForType):
(WebCore::RenderSVGPath::shouldGenerateMarkerPositions const):
(WebCore::RenderSVGPath::drawMarkers):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
(WebCore::RenderSVGPath::processMarkerPositions):
* Source/WebCore/rendering/svg/RenderSVGPath.h:
* Source/WebCore/rendering/svg/RenderSVGRect.cpp:
(WebCore::RenderSVGRect::updateShapeFromElement):
(WebCore::RenderSVGRect::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::updateShapeFromElement):
(WebCore::RenderSVGShape::fillStrokeMarkers):
(WebCore::RenderSVGShape::ensurePath):
(WebCore::RenderSVGShape::shouldGenerateMarkerPositions const): Deleted.
(WebCore::markerForType): Deleted.
(WebCore::RenderSVGShape::computeMarkerBoundingBox const): Deleted.
(WebCore::RenderSVGShape::drawMarkers): Deleted.
(WebCore::RenderSVGShape::processMarkerPositions): Deleted.
* Source/WebCore/rendering/svg/RenderSVGShape.h:
(WebCore::RenderSVGShape::shapeType const):
(WebCore::RenderSVGShape::drawMarkers):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleShapeOrTextOrInline const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGEllipse.cpp:
(WebCore::LegacyRenderSVGEllipse::updateShapeFromElement):
(WebCore::LegacyRenderSVGEllipse::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::updateShapeFromElement):
(WebCore::LegacyRenderSVGPath::calculateUpdatedStrokeBoundingBox const):
(WebCore::markerForType):
(WebCore::LegacyRenderSVGPath::shouldGenerateMarkerPositions const):
(WebCore::LegacyRenderSVGPath::drawMarkers):
(WebCore::LegacyRenderSVGPath::markerRect const):
(WebCore::LegacyRenderSVGPath::processMarkerPositions):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRect.cpp:
(WebCore::LegacyRenderSVGRect::updateShapeFromElement):
(WebCore::LegacyRenderSVGRect::shapeDependentStrokeContains):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::updateShapeFromElement):
(WebCore::LegacyRenderSVGShape::layout):
(WebCore::LegacyRenderSVGShape::fillStrokeMarkers):
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):
(WebCore::LegacyRenderSVGShape::ensurePath):
(WebCore::LegacyRenderSVGShape::shouldGenerateMarkerPositions const): Deleted.
(WebCore::markerForType): Deleted.
(WebCore::LegacyRenderSVGShape::markerRect const): Deleted.
(WebCore::LegacyRenderSVGShape::drawMarkers): Deleted.
(WebCore::LegacyRenderSVGShape::processMarkerPositions): Deleted.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.h:
(WebCore::LegacyRenderSVGShape::drawMarkers):

Canonical link: https://commits.webkit.org/269245@main
